### ISSUE TITLE
[H5] The aggregation path: cross-attention head vs a fine-tuned head control

### ISSUE THREAD

**Issue by Parswanadh** [Parswanadh/laya-lab#6]
Owner: `engineer`. Charter `roles/README.md` §engineer. **Take the GPU flock.**

## Why this, and why now
`findings/P1b-information-location.md` measured that the frozen encoder delivers label-recoverable information to the option-marker positions at **0.820** accuracy (linear probe, **held-out templates**, needle 4000 tokens away) while the **shipped head returns 0.300** on the same items.

Encoder attention surgery is therefore not where the headroom is — which independently matches R-001's published prior (8× window → +0.4 micro-F1).

## The confound you must control for — this is the point of the task
A probe fitted on the evaluation distribution is **not** a fair comparison to a frozen head trained on a different task. So a win for a new head could be **training**, not **architecture**. Without arm 2 below, the whole result is uninterpretable.

## Arms (all with the encoder FROZEN, all on identical data and identical eval items)

| arm | architecture | isolates |
|---|---|---|
| 1 | shipped head + scorer, frozen | baseline (0.300 measured at needle@END) |
| 2 | **same head architecture, fine-tuned** | **is fine-tuning alone sufficient?** |
| 3 | **cross-attention head** over all encoded state positions, ~matched parameter count to arm 2's trainable set | does the *architecture* add anything over arm 2? |
| 4 *(stretch)* | arm 3 + oversampled long-distance examples | is any remaining gap a data-distribution artefact? |

**If arm 2 ≈ arm 3, the architecture is not the win** and the honest conclusion is that the fix is adaptation, not design. That outcome is pre-registered and is a perfectly good result — report it as clearly as a win.

## Arm 3 specification
Replace the decision head's self-attention over the flat sequence with a **learned-query cross-attention** over the encoder's state representations:

```
Stage 1  encoder (FROZEN)  D -> H_D in R^(n x 768)          full bidirectional, native 8192
Stage 2  decision tokens   [CLS] type instr [SEP] [MASK] o1 .. ok [SEP] -> H_M in R^(m x 768)
Stage 3  CROSS-ATTENTION (NEW)  Q=W_q H_M ; K=W_k H_D ; V=W_v H_D
         H_M <- H_M + MHA(Q,K,V) + FFN(.)                   ~2 layers, 4 heads
Stage 4  scoring (UNCHANGED)  gather at marker positions -> scorer -> logits
```
Every state token is a K/V pair, so each marker reaches **all n positions in one hop** with a query **trained for this task**, instead of routing through the 8 global layers while competing with every filler token for attention mass. Marginal cost is ~1–2% of the encoder's (modelled; measure it).

Reuse `laya/common.py:DecisionModel` and `build_sequence` — do **not** fork the model code into a parallel universe. Keep the arm selectable so upstream's default path is untouched.

## Data — leakage rules are strict
- Needles generated from **templates disjoint from every evaluation template**, verified **by content hash**, not by variable name. `experiments/harness/pools/needles-balanced-v1.json` exists; extend it or author a new pool, but the train/eval split must be hash-verified and the check committed.
- **Balanced labels** (4 classes equal). Report the label counts.
- Vary needle position and document length across the training set, or arm 3 will only learn one regime. State the training length/position distribution explicitly.
- The cross-verifier will re-derive your split and attempt to falsify it.

## Evaluation — follow `protocol.md` exactly
Held-out templates · balanced · needle position swept at **fixed** document length · `n >= 200` for any number you quote · report **random (0.25), majority, and the POSITION-ONLY ORACLE** for every cell · bootstrap CIs · **McNemar** paired vs arm 1 and vs arm 2 · seed spread over >=3 seeds.

**The position-only oracle is mandatory on this arm.** An adapter that learns 'needle near the end implies billing' will look excellent without reading anything. If arm 3 does not beat the position-only oracle, say so in exactly those words.

## Feasibility (verified on this box)
Encoder frozen under `torch.no_grad()` + `detach()`, batch 2–4 at 4096 tokens, gradients only through the ~3M-param adapter — fits the 8 GB card. Confirm with `free -h` and a VRAM print before the long run. Cap memory; log inside the repo.

## Deliverables
- Training + eval code on a branch under `worktrees/` (never on `main`).
- `experiments/h5-adapter/manifest.json` + raw per-item JSONL for every arm.
- `findings/E-003.md` — the four-arm table, the leakage check, the oracle baselines, and an explicit verdict: **does the architecture beat the fine-tuned control?** Include the negative outcome if that is what the data says.
- A runnable check left behind (upstream `AGENTS.md` requires this for non-trivial logic).

## Do not
- Do not modify `fork/laya/common.py` on `main`; work on `exp/h5-adapter`.
- Do not grade your own work — report numbers, verification assigns confidence.
- Do not report a number without the `n` and the artifact path behind it.
